roslibrust/src/ros1/master_client.rs

```diff
@@ -0,0 +1,360 @@
+    /// - id: A client_id to use when communicating with the master, expected to be a valid ros name e.g. "/my_node"
+    pub async fn new(
+        master_uri: impl Into<String>,
+        client_uri: impl Into<String>,
+        id: impl Into<String>,
+    ) -> Result<MasterClient, RosMasterError> {
+        // Create a client, but we want to verify a valid connection before handing control back,
+        // so we make an initial request and confirm with works before returning
+        let client = MasterClient {
+            client: reqwest::Client::new(),
+            master_uri: master_uri.into(),
+            client_uri: client_uri.into(),
+            id: id.into(),
+        };
+
+        match client.get_uri().await {
+            Ok(_) => Ok(client),
+            Err(e) => Err(e),
+        }
+    }
+
+    async fn post<T: serde::de::DeserializeOwned + std::fmt::Debug>(
+        &self,
+        request: String,
+    ) -> Result<T, RosMasterError> {
+        trace!("Sending master: {request}");
+        let response = self
+            .client
+            .post(&self.master_uri)
+            .body(request)
+            .send()
+            .await?
+            .text()
+            .await?;
+        trace!("Got response: {response}");
+        let parsed: (i8, String, T) = serde_xmlrpc::response_from_str(&response)?;
+        match parsed.0 {
+            1 => {
+                trace!("Parsed from rosmaster: {:?}", parsed.2);
+            }
+            _ => {
+                return Err(RosMasterError::MasterError(parsed.1));
+            }
+        };
+        Ok(parsed.2)
+    }
+
+    /// Hits the master's xmlrpc endpoint "getUri" and provides the response
+    pub async fn get_uri(&self) -> Result<String, RosMasterError> {
+        let body = serde_xmlrpc::request_to_string("getUri", vec![self.id.clone().into()])?;
+        self.post(body).await
+    }
+
+    /// Hits the master's xmlrpc endpoint "getTopicTypes" and provides the response as
+    /// Vec<(topicName, topicType)>
+    pub async fn get_topic_types(&self) -> Result<Vec<(String, String)>, RosMasterError> {
+        let body = serde_xmlrpc::request_to_string("getTopicTypes", vec![self.id.clone().into()])?;
+        self.post(body).await
+    }
+
+    /// Hits the master's xmlrpc endpoint "registerService", does not actually begin hosting the service uri.
+    pub async fn register_service(
+        &self,
+        service: impl Into<String>,
+        service_uri: impl Into<String>,
+    ) -> Result<(), RosMasterError> {
+        let body = serde_xmlrpc::request_to_string(
+            "registerService",
+            vec![
+                self.id.clone().into(),
+                service.into().into(),
+                service_uri.into().into(),
+                self.client_uri.clone().into(),
+            ],
+        )?;
+        // Little conversion here to ignore third response parameter
+        let _: u8 = self.post(body).await?;
```

Comment:
Why specify the type for something that is being explicitly ignored?

Reply:
Actually have to for the templates to work out. ROS is returning a u8, but the api docs literally say that it should be ignored. If we don't parse the u8 the deserialization fails.

I think `self.post<u8>(body)` also works, but I kinda like this style?